File: mobile/ios/KompaniyaAnticor/Networking/APIClient.swift
import Foundation

/// Backendning yagona action-dispatch endpoint'i bilan gaplashadigan past
/// darajadagi klient. Har bir chaqiruv `POST {baseURL}/index.php`ga
/// `{"action": "...", ...}` JSON tanasi bilan boradi.
///
/// Autentifikatsiya: backend HttpOnly cookie'dan tashqari
/// "Authorization: Bearer <token>" sarlavhasini ham qabul qiladi (qarang:
/// backend/src/Auth.php::tokenFromRequest()). Token doim Keychain orqali
/// saqlanadi — hech qachon UserDefaults yoki oddiy xotirada emas.
final class APIClient {
    /// Backend'ning ochiq manzili — server joylashuviga qarab o'zgartiring.
    /// Mahalliy rivojlantirishda: iOS simulyatorida "http://localhost:8080/backend/public"
    /// ishlaydi (simulyator xost kompyuterning localhost'ini to'g'ridan-to'g'ri ko'radi);
    /// haqiqiy qurilmada kompyuteringizning lokal tarmoq IP manzilini yozing.
    static let defaultBaseURL = "https://your-domain.uz/backend/public"

    let baseURL: String
    private let session: URLSession
    private static let tokenKey = "session_token"
    private static let rememberTokenKey = "remember_token"

    init(baseURL: String = APIClient.defaultBaseURL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    var storedToken: String? { KeychainStore.get(Self.tokenKey) }
    func saveToken(_ token: String) { KeychainStore.set(token, forKey: Self.tokenKey) }
    func clearToken() { KeychainStore.delete(Self.tokenKey) }

    /// "Meni eslab qol" tokeni — sessiya harakatsizlikdan (10 daqiqa) tugab
    /// qolganda parolsiz jim qayta kirish uchun (mobileLoginViaRememberToken).
    /// Mutlaq 1 soatlik umr chegarasiga ega, hech qachon uzaytirilmaydi.
    var storedRememberToken: String? { KeychainStore.get(Self.rememberTokenKey) }
    func saveRememberToken(_ token: String) { KeychainStore.set(token, forKey: Self.rememberTokenKey) }
    func clearRememberToken() { KeychainStore.delete(Self.rememberTokenKey) }

    /// Asosiy chaqiruv usuli. `params` ichiga `action`ni qo'shmang — u
    /// alohida beriladi. Muvaffaqiyatli javobning to'liq xaritasini
    /// qaytaradi; xatolikda APIError otadi.
    @discardableResult
    func call(_ action: String, _ params: [String: Any] = [:]) async throws -> [String: Any] {
        guard let url = URL(string: "\(baseURL)/index.php") else {
            throw APIError(code: "BAD_URL", message: "Noto'g'ri server manzili", statusCode: 0)
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = storedToken, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        var body = params
        body["action"] = action
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError(code: "NETWORK_ERROR", message: "Tarmoq xatoligi: so'rov yuborilmadi", statusCode: 0)
        }

        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw APIError(code: "BAD_RESPONSE", message: "Server javobini o'qib bo'lmadi", statusCode: statusCode)
        }

        if (json["success"] as? Bool) != true {
            throw APIError(
                code: json["code"] as? String ?? "ERROR",
                message: json["message"] as? String ?? "Noma'lum xatolik",
                statusCode: statusCode
            )
        }
        return json
    }

    /// document-download.php / hr-document-download.php / purchase-download.php
    /// / backup-download.php kabi JSON action-dispatch'dan tashqaridagi GET
    /// skriptlarni Bearer sarlavhasi bilan yuklab oladi. [pathAndQuery]
    /// `baseURL`dan keyingi qism, masalan "document-download.php?id=5".
    func downloadFile(_ pathAndQuery: String) async throws -> Data {
        guard let url = URL(string: "\(baseURL)/\(pathAndQuery)") else {
            throw APIError(code: "BAD_URL", message: "Noto'g'ri server manzili", statusCode: 0)
        }
        var request = URLRequest(url: url)
        if let token = storedToken, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError(code: "NETWORK_ERROR", message: "Tarmoq xatoligi: fayl yuklanmadi", statusCode: 0)
        }
        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard statusCode == 200 else {
            throw APIError(code: "DOWNLOAD_FAILED", message: "Faylni yuklab bo'lmadi", statusCode: statusCode)
        }
        return data
    }

    /// Binary faylni backend kutgan "data:<mime>;base64,..." shakliga o'giradi
    /// (hujjat/rasm yuklash amallari — addDocument, submitHrDocument,
    /// addPurchase, updateProfilePhoto — shu formatni JSON tana ichida kutadi).
    static func toDataURL(_ data: Data, mimeType: String) -> String {
        "data:\(mimeType);base64,\(data.base64EncodedString())"
    }
}

/// `[String: Any]` javob xaritasidan qulay tipdagi qiymat olish uchun kichik yordamchilar.
extension Dictionary where Key == String, Value == Any {
    func str(_ key: String) -> String? { self[key] as? String }
    func int(_ key: String) -> Int? {
        if let v = self[key] as? Int { return v }
        if let v = self[key] as? NSNumber { return v.intValue }
        return nil
    }
    func bool(_ key: String) -> Bool? { self[key] as? Bool }
    func double(_ key: String) -> Double? {
        if let v = self[key] as? Double { return v }
        if let v = self[key] as? NSNumber { return v.doubleValue }
        if let v = self[key] as? String { return Double(v) }
        return nil
    }
    func arr(_ key: String) -> [[String: Any]] {
        (self[key] as? [[String: Any]]) ?? []
    }
}

File: mobile/flutter/lib/api/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import 'api_exception.dart';

/// Backendning yagona action-dispatch endpoint'i bilan gaplashadigan
/// past darajadagi klient. Har bir chaqiruv `POST {baseUrl}/index.php`ga
/// `{"action": "...", ...}` JSON tanasi bilan boradi — xuddi veb frontend
/// (main.html/admin.html) ishlatadigan `apiCall()` bilan bir xil andoza.
///
/// Autentifikatsiya: backend HttpOnly cookie'dan tashqari
/// "Authorization: Bearer <token>" sarlavhasini ham qabul qiladi (qarang:
/// backend/src/Auth.php::tokenFromRequest()). Token doim
/// flutter_secure_storage orqali (iOS Keychain / Android Keystore)
/// saqlanadi — hech qachon oddiy SharedPreferences yoki xotirada oddiy
/// o'zgaruvchida emas.
class ApiClient {
  /// Backend'ning ochiq manzili — server joylashuviga qarab o'zgartiring.
  /// Mahalliy rivojlantirishda masalan: http://10.0.2.2:8080/backend/public
  /// (Android emulyatori "localhost"ni shunday ko'radi) yoki
  /// http://localhost:8080/backend/public (iOS simulyatori uchun).
  static const String defaultBaseUrl = 'https://your-domain.uz/backend/public';

  final String baseUrl;
  final http.Client _http;
  final FlutterSecureStorage _storage;

  static const _tokenKey = 'session_token';
  static const _rememberTokenKey = 'remember_token';

  ApiClient({String? baseUrl, http.Client? httpClient, FlutterSecureStorage? storage})
      : baseUrl = baseUrl ?? defaultBaseUrl,
        _http = httpClient ?? http.Client(),
        _storage = storage ?? const FlutterSecureStorage();

  Future<String?> get storedToken => _storage.read(key: _tokenKey);

  Future<void> saveToken(String token) => _storage.write(key: _tokenKey, value: token);

  Future<void> clearToken() => _storage.delete(key: _tokenKey);

  /// "Meni eslab qol" tokeni — sessiya harakatsizlikdan (10 daqiqa) tugab
  /// qolganda parolsiz jim qayta kirish uchun (mobileLoginViaRememberToken).
  /// Mutlaq 1 soatlik umr chegarasiga ega, hech qachon uzaytirilmaydi.
  Future<String?> get storedRememberToken => _storage.read(key: _rememberTokenKey);

  Future<void> saveRememberToken(String token) => _storage.write(key: _rememberTokenKey, value: token);

  Future<void> clearRememberToken() => _storage.delete(key: _rememberTokenKey);

  /// Asosiy chaqiruv usuli. `params` ichiga `action`ni qo'shmang — u
  /// alohida beriladi. Muvaffaqiyatli javobning to'liq (`success` ham
  /// ichida bo'lgan) xaritasini qaytaradi; xatolikda ApiException otadi.
  Future<Map<String, dynamic>> call(String action, [Map<String, dynamic> params = const {}]) async {
    final token = await storedToken;
    final headers = <String, String>{
      'Content-Type': 'application/json',
      if (token != null && token.isNotEmpty) 'Authorization': 'Bearer $token',
    };

    final body = jsonEncode({'action': action, ...params});

    late http.Response resp;
    try {
      resp = await _http.post(Uri.parse('$baseUrl/index.php'), headers: headers, body: body);
    } catch (e) {
      throw ApiException(code: 'NETWORK_ERROR', message: "Tarmoq xatoligi: so'rov yuborilmadi", statusCode: 0);
    }

    Map<String, dynamic> data;
    try {
      data = jsonDecode(resp.body) as Map<String, dynamic>;
    } catch (e) {
      throw ApiException(code: 'BAD_RESPONSE', message: "Server javobini o'qib bo'lmadi", statusCode: resp.statusCode);
    }

    if (data['success'] != true) {
      throw ApiException(
        code: (data['code'] as String?) ?? 'ERROR',
        message: (data['message'] as String?) ?? "Noma'lum xatolik",
        statusCode: resp.statusCode,
      );
    }

    return data;
  }

  /// document-download.php / hr-document-download.php / purchase-download.php
  /// / backup-download.php kabi JSON action-dispatch'dan tashqaridagi
  /// GET skriptlarni Bearer sarlavhasi bilan yuklab oladi. [pathAndQuery]
  /// `baseUrl`dan keyingi qism, masalan "document-download.php?id=5".
  Future<List<int>> downloadFile(String pathAndQuery) async {
    final token = await storedToken;
    final headers = <String, String>{
      if (token != null && token.isNotEmpty) 'Authorization': 'Bearer $token',
    };
    late http.Response resp;
    try {
      resp = await _http.get(Uri.parse('$baseUrl/$pathAndQuery'), headers: headers);
    } catch (e) {
      throw ApiException(code: 'NETWORK_ERROR', message: "Tarmoq xatoligi: fayl yuklanmadi", statusCode: 0);
    }
    if (resp.statusCode != 200) {
      throw ApiException(code: 'DOWNLOAD_FAILED', message: "Faylni yuklab bo'lmadi", statusCode: resp.statusCode);
    }
    return resp.bodyBytes;
  }

  /// Binary faylni backend kutgan "data:<mime>;base64,..." shakliga o'giradi
  /// (hujjat/rasm yuklash amallari — addDocument, submitHrDocument,
  /// addPurchase, updateProfilePhoto — shu formatni JSON tana ichida kutadi).
  static String toDataUrl(List<int> bytes, String mimeType) {
    return 'data:$mimeType;base64,${base64Encode(bytes)}';
  }
}
